keps/sig-storage/3141-prevent-volume-mode-conversion/README.md

```diff
@@ -0,0 +1,820 @@
+
+## Summary
+
+With the introduction of the VolumeSnapshot feature in Kubernetes 1.20, a
+PersistentVolumeClaim (PVC) can be restored from a previously taken VolumeSnapshot.
+This is done by pointing the “dataSource” parameter of the PVC spec to an existing
+VolumeSnapshot resource. There is no logic that validates whether the original
+volume mode of the PVC, whose snapshot was taken, matches the volume mode of 
+the newly created PVC that is being restored from the existing VolumeSnapshot.
+This KEP proposes a solution to prevent unauthorized conversion of the volume
+mode during a restore operation. 
+
+## Motivation
+
+Malicious users may expose a vulnerability in the kernel by exploiting
+this gap. 
+Here is an example of how a malicious user can exploit this gap to crash the
+kernel.
+1. User creates a PVC with “volumeMode: Block” and runs a pod with it.
+2. User writes malformed ext4 data to it (simple dd)
+3. User takes snapshot of this volume.
+4. User restores snapshot as a PVC with “volumeMode: Filesystem”
+5. User uses this PVC in a pod.
+    1. kubelet tries to mount it during pod creation. If there is a CVE in the
+       kernel, the user can crash it.
+
+However we cannot simply block this operation as some backup
+vendors try to restore a volume with the exact same mode as the original
+volume but may need to do the conversion for efficiency.
+
+### Goals
+
+Define a mechanism to mitigate the vulnerability of restoring volumes without
+hampering valid use cases.
+
+### Non-Goals
+
+Design that is generic and can be extended to other storage related security
+aspects. This proposal only addresses the snapshot restore security issue.
+
+## Proposal
+
+The proposal aims to mitigate this issue by modifying the `VolumeSnapshotContent`
+API spec as well as the control flows of `snapshot-controller` and `external-provisioner`.
+`VolumeSnapshotContent` API will include a field that denotes the volume mode of
+the volume that the snapshot was created from. 
+This proposal also introduces a new annotation on the `VolumeSnapshotContent` resource
+that a trusted user (like a backup software) needs to apply on a VolumeSnapshot.
+
+By introducing these changes, we will leverage existing user access rights to determine
+whether the volume mode of a volume can be altered during a snapshot restore operation.
+
+### User Stories (Optional)
+
+#### Story 1
+
+When a `VolumeSnapshot` is created from an existing `PVC`, a corresponding 
+`VolumeSnapshotContent` is created by the `snapshot-controller`.
+Alternatively, a `VolumeSnapshotContent` can be manually created by an admin
+if the `Spec.Source.SnapshotHandle` refers to a pre-existing snapshot on the
+underlying storage system. 
+
+In either case, `VolumeSnapshots` and `VolumeSnapshotContents` maintain a 1:1 
+mapping. Backup vendors that need to convert the volume mode during a restore
+operation need to identify the `VolumeSnapshotContent` mapped to the
```

Comment:
The conversion actually happens during a backup operation.
1. Assume the original PVC is created with filesystem mode. 
2. During backup, the backup software creates a 2nd PVC from a VolumeSnapshot with block mode. This needs volume mode conversion. The purpose of creating this 2nd PVC with block mode is to be able to copy data efficiently and save it to a backup target.
3. The 2nd PVC is temporary and will be deleted after data is copied. 
4. At restore time, a new PVC (the 3rd PVC) will be created with the original filesystem mode.

Reply:
Thanks! I've updated the `User Story` as well as `Motivation` sections with this information.